src/rust/src/pkcs7.rs

```diff
@@ -164,6 +166,257 @@ fn encrypt_and_serialize<'p>(
+                        pyo3::exceptions::PyValueError::new_err(
+                            "No recipient found that matches the given certificate.",
+                        ),
+                    ));
+                }
+            };
+
+            // Raise error when the key encryption algorithm is not RSA
+            let key = match recipient_info.key_encryption_algorithm.oid() {
+                &oid::RSA_OID => {
+                    let padding = types::PKCS1V15.get(py)?.call0()?;
+                    private_key
+                        .call_method1(
+                            pyo3::intern!(py, "decrypt"),
+                            (recipient_info.encrypted_key, &padding),
+                        )?
+                        .extract::<pyo3::pybacked::PyBackedBytes>()?
+                }
+                _ => {
+                    return Err(CryptographyError::from(
+                        exceptions::UnsupportedAlgorithm::new_err((
+                            "Only RSA with PKCS #1 v1.5 padding is currently supported for key decryption.",
+                            exceptions::Reasons::UNSUPPORTED_SERIALIZATION,
+                        )),
+                    ));
+                }
+            };
+
+            // Get algorithm
+            // TODO: implement all the possible algorithms
+            let algorithm_identifier = enveloped_data
+                .encrypted_content_info
+                .content_encryption_algorithm;
+            let (algorithm, mode) = match algorithm_identifier.params {
+                AlgorithmParameters::Aes128Cbc(iv) => (
+                    types::AES128.get(py)?.call1((key,))?,
+                    types::CBC
+                        .get(py)?
+                        .call1((pyo3::types::PyBytes::new(py, &iv),))?,
+                ),
+                _ => {
+                    return Err(CryptographyError::from(
+                        exceptions::UnsupportedAlgorithm::new_err((
+                            "Only AES-128-CBC is currently supported for content decryption.",
+                            exceptions::Reasons::UNSUPPORTED_SERIALIZATION,
+                        )),
+                    ));
+                }
+            };
+
+            // Decrypt the content using the key and proper algorithm
+            let encrypted_content = enveloped_data
+                .encrypted_content_info
+                .encrypted_content
+                .unwrap();
+            let decrypted_content = symmetric_decrypt(py, algorithm, mode, encrypted_content)?;
+            pyo3::types::PyBytes::new(py, decrypted_content.as_slice())
+        }
+        _ => {
+            return Err(CryptographyError::from(
+                pyo3::exceptions::PyValueError::new_err(
+                    "The PKCS7 data is not an EnvelopedData structure.",
+                ),
+            ));
+        }
+    };
+
+    // If text_mode, strip the header if possible, else return an error
```

Comment:
@reaperhulk can I get your opinion on the right behavior for text vs. binary?

Reply:
We added these flags to behave like OpenSSL (sigh).

OpenSSL calls `SMIME_text` in its decryption path when you pass `PKCS7_TEXT`. That function errors for a few conditions, but most notably if `parsing the MIME headers fails, if there is no “content-type” header, or if the content type is not “text/plain”.`

This code appears to make two assumptions that may not match OpenSSL's though:
* Case sensitivity on `Content-Type`
* That the Content-Type header is guaranteed to be the last header before the payload.

The behavior I believe we want is to parse the MIME headers, check that content-type is text/plain, and then discard them entirely and only keep the payload.

Reply:
Do you have any suggestion of a lib to handle MIME efficiently in rust? For now, I'm thinking about
using some Python function. Seems simpler to me.

Reply:
I would just use Python for this.

Reply:
First version pushed, will compare with `openssl` and let you know!

And yeah, it should be up to the user to sanitize the decrypted output after `pkcs7_decrypt`, especially with the libraries available on Python that are perfect for this ...

Reply:
I've tested against some custom headers examples, behavior is the same as `openssl` (using `test_support.pkcs7_decrypt` from `main`).